### ISSUE TITLE
[Enhancement] Add ability to select CPU?

### ISSUE THREAD

**Issue by computernewb** [copy/v86#85]
I think it'd be really cool if this site let you choose which CPU it emulated. I made a concept of what I mean:

<img src="http://i.imgur.com/sU4RoH6.png"></img>

I would be willing to help you code the emulator for the CPUs in question, if you'd like.

**Comment by copy** [comment]
> I would be willing to help you code the emulator for the CPUs in question, if you'd like. 

Sounds good. Do you know an OS that requires specific CPU models?

What do we need to change internally to do this?

**Comment by computernewb** [comment]
I have not taken a full look at the code yet so I am unsure of what you would need to change internally. 

I believe Windows NT 3.1 require specific CPUs to run properly otherwise it causes a panic in the kernel and immediately quits the setup. 

Certain games also do not run properly, as they were designed for a specific processor.

**Comment by Creeper20428** [comment]
I like it too. But I think that having a choice for people to select it on the site is a bad idea. Let me explain.

The obvious solution would be to surround certain operations with `if` statements:
`if(CPU8086 || CPU80186){ ... }else{ /* for 80286+ */ ... }`
When compiling with closure, it's simple enough to do this:
`/** @const */var CPU8086 = false;`
Then, closure inlines the variable, optimizes the if statement accordingly, and we end up with no branching all!

But, if users are allowed to choose what cpu they on the web site, this would have to happen:
`if(cpu.is8086){ ... }else{ ... }`
That branch can't be removed, and that condition has to be evaluated every single time that instruction executes, slowing down execution (and I'm aware of what happens when a branch misprediction happens. Ugh)

In conclusion, I like the idea -- but making that a configurable choice at run time would make performance break down. 
Maybe we could have a simple "./configure" script that automagically switches around variables before a compile.

Another idea I had just now was that the user could select CPUs from the web site -- and we conditionally loaded the files on the fly (e.g. "v86-80386.js" for a 386 CPU). That seems (somewhat) feasible...

**Comment by copy** [comment]
> That branch can't be removed, and that condition has to be evaluated every single time that instruction executes, slowing down execution

I agree with this. We can also reduce the impact on instructions by overwriting entries in the instruction table, for example

```js
if(cpu.is8086)
{
	t[0x0f] = cpu => { /* pop cs */ };
}
else
{
	t[0x0f] = cpu => { /*  0f prefix */ };
}
```

> Maybe we could have a simple "./configure" script that automagically switches around variables before a compile.

I'm not a fan of more complicated build steps. This can also be done with Closure Compiler (if we really want to do this):

`if(RUNTIME_CPU_CONFIG ? cpu.is8086 : CPU8086) …`

and Closure Compiler will simplify the expression accordingly.

**Comment by Creeper20428** [comment]
I don't like fancy `./configure` scripts that take like ten minutes to run on my computer (especially since I have Windows). But that was a suggestion that I just threw out.

I just remembered that the 8086 has weird handling for undefined opcodes. http://www.os2museum.com/wp/undocumented-8086-opcodes/
That could be fun to implement.

**Comment by Creeper20428** [comment]
How about this (just thought of this a few minutes ago):
```javascript
function setCPU_Pentium(){
    op16[0x00] = cpu => { /* ADD Eb, Gb */ };
    // etc. etc.
}
function setCPU_8086(){
    op16[0x0F] = cpu => { /* POP CS */ };
    op16[0x10] = cpu => { /* OR Eb, Gb */ };
}
```
That way, there won't be an need for separate files for different CPUs.

**Comment by copy** [comment]
@Creeper20428 Yep, that'd work (I suggested something similar above).

**Comment by Creeper20428** [comment]
The only problem that I can see is that it would bloat the file size a lot with redundant operations.
